
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Graph-based linkage disequilibrium matrix operations and heritability estimation">
      
      
      
        <link rel="canonical" href="https://oclb.github.io/graphld/cli/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../score_test/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Command Line Interface - GraphLD v1.2.0</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#command-line-interface" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GraphLD v1.2.0" class="md-header__button md-logo" aria-label="GraphLD v1.2.0" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GraphLD v1.2.0
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Command Line Interface
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/oclb/graphld" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    oclb/graphld
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  
  User Guide

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../api/graphld/" class="md-tabs__link">
          
  
  
  API Reference

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../see_also/" class="md-tabs__link">
        
  
  
    
  
  See Also

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GraphLD v1.2.0" class="md-nav__button md-logo" aria-label="GraphLD v1.2.0" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    GraphLD v1.2.0
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oclb/graphld" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    oclb/graphld
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Installation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    User Guide
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    User Guide
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Command Line Interface
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Command Line Interface
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#available-commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        Available Commands
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#heritability-estimation-reml" class="md-nav__link">
    <span class="md-ellipsis">
      
        Heritability Estimation (reml)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Heritability Estimation (reml)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input Formats
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output Files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#options-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Options Reference
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-set-annotations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Gene Set Annotations
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#surrogate-markers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Surrogate Markers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parquet-multi-trait-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parquet Multi-Trait Files
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#blup-weights-blup" class="md-nav__link">
    <span class="md-ellipsis">
      
        BLUP Weights (blup)
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ld-clumping-clump" class="md-nav__link">
    <span class="md-ellipsis">
      
        LD Clumping (clump)
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulation-simulate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Simulation (simulate)
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../score_test/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Enrichment Score Test
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../file_formats/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    File Formats
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    API Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    API Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/graphld/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/precision/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.precision
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/likelihood/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.likelihood
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/heritability/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.heritability
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/ldsc_io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.ldsc_io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/vcf_io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.vcf_io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/parquet_io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.parquet_io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/simulate/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.simulate
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/clumping/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.clumping
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/blup/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.blup
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/surrogates/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.surrogates
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/genesets/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.genesets
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/multiprocessing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    graphld.multiprocessing_template
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/score_test/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    score_test
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../see_also/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    See Also
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#available-commands" class="md-nav__link">
    <span class="md-ellipsis">
      
        Available Commands
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#heritability-estimation-reml" class="md-nav__link">
    <span class="md-ellipsis">
      
        Heritability Estimation (reml)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Heritability Estimation (reml)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input Formats
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output Files
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#options-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Options Reference
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-set-annotations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Gene Set Annotations
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#surrogate-markers" class="md-nav__link">
    <span class="md-ellipsis">
      
        Surrogate Markers
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parquet-multi-trait-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parquet Multi-Trait Files
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#blup-weights-blup" class="md-nav__link">
    <span class="md-ellipsis">
      
        BLUP Weights (blup)
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ld-clumping-clump" class="md-nav__link">
    <span class="md-ellipsis">
      
        LD Clumping (clump)
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#simulation-simulate" class="md-nav__link">
    <span class="md-ellipsis">
      
        Simulation (simulate)
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="command-line-interface">Command Line Interface<a class="headerlink" href="#command-line-interface" title="Permanent link">&para;</a></h1>
<p>The GraphLD CLI provides commands for various LD-based analyses. Use <code>uv run graphld &lt;command&gt; -h</code> to see usage for a specific command.</p>
<p>For the enrichment score test CLI (<code>estest</code>), see <a href="../score_test/">Enrichment Score Test</a>.</p>
<h2 id="available-commands">Available Commands<a class="headerlink" href="#available-commands" title="Permanent link">&para;</a></h2>
<table>
<thead>
<tr>
<th>Command</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>reml</code></td>
<td>Run graphREML heritability estimation</td>
</tr>
<tr>
<td><code>blup</code></td>
<td>Compute best linear unbiased predictor weights</td>
</tr>
<tr>
<td><code>clump</code></td>
<td>P-value thresholding and LD-based pruning</td>
</tr>
<tr>
<td><code>simulate</code></td>
<td>Simulate GWAS summary statistics</td>
</tr>
<tr>
<td><code>surrogates</code></td>
<td>Precompute surrogate markers</td>
</tr>
</tbody>
</table>
<h2 id="heritability-estimation-reml">Heritability Estimation (reml)<a class="headerlink" href="#heritability-estimation-reml" title="Permanent link">&para;</a></h2>
<p>Run graphREML for heritability partitioning:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>reml<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w">    </span>/path/to/sumstats/file.sumstats<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w">    </span>output_files_prefix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="w">    </span>--annot-dir<span class="w"> </span>/directory/containing/annotation/files/
</code></pre></div>
<h3 id="input-formats">Input Formats<a class="headerlink" href="#input-formats" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>Summary statistics</strong>: VCF (<code>.vcf</code>), LDSC (<code>.sumstats</code>), or Parquet (<code>.parquet</code>) format</li>
<li><strong>Variant annotations</strong>: Per-chromosome <a href="https://github.com/bulik/ldsc/wiki/LD-File-Formats#annot">LDSC <code>.annot</code> files</a>, including <code>thin-annot</code> format</li>
<li><strong>BED files</strong>: UCSC <code>.bed</code> files (not stratified per-chromosome) for binary annotations based on GRCh38 coordinates</li>
<li><strong>Gene annotations</strong>: GMT files (<code>.gmt</code>) containing gene sets, converted to variant-level using nearest-gene weighting</li>
</ul>
<h3 id="output-files">Output Files<a class="headerlink" href="#output-files" title="Permanent link">&para;</a></h3>
<p><strong>Default output:</strong></p>
<ul>
<li><code>output_prefix.tall.csv</code>: Heritability, enrichment, and coefficient estimates for each annotation</li>
<li><code>output_prefix.convergence.csv</code>: Information about the optimization process</li>
</ul>
<p><strong>With <code>--alt-output</code> flag</strong> (useful for storing results from multiple traits or runs in a single file):</p>
<ul>
<li><code>output_prefix.heritability.csv</code>: Heritability estimates (one row per run)</li>
<li><code>output_prefix.enrichment.csv</code>: Enrichment estimates (one row per run)</li>
<li><code>output_prefix.parameters.csv</code>: Coefficient estimates (one row per run)</li>
</ul>
<p>Use <code>--name</code> to label each run when appending to these files.</p>
<h3 id="options-reference">Options Reference<a class="headerlink" href="#options-reference" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>uv run graphld reml --help
</code></pre></div>
<p><strong>Required (one of):</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sumstats</code></td>
<td>Path to summary statistics file (.vcf, .sumstats, or .parquet)</td>
</tr>
<tr>
<td><code>-a, --annot-dir</code></td>
<td>Path to annotation directory containing .annot and/or .bed files</td>
</tr>
<tr>
<td><code>-g, --gene-annot-dir</code></td>
<td>Path to directory containing .gmt gene set files</td>
</tr>
</tbody>
</table>
<p><strong>Common options:</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>out</code></td>
<td>None</td>
<td>Output file path prefix</td>
</tr>
<tr>
<td><code>--intercept</code></td>
<td>1.0</td>
<td>LD score regression intercept (recommended to estimate with LDSC first)</td>
</tr>
<tr>
<td><code>-n, --num-samples</code></td>
<td>auto</td>
<td>Sample size (auto-detected from sumstats if available)</td>
</tr>
<tr>
<td><code>-p, --population</code></td>
<td>EUR</td>
<td>Population for LDGM selection</td>
</tr>
<tr>
<td><code>-c, --chromosome</code></td>
<td>all</td>
<td>Restrict analysis to specific chromosome</td>
</tr>
<tr>
<td><code>--name</code></td>
<td>None</td>
<td>Label for this run (used in alt-output files and score test HDF5)</td>
</tr>
<tr>
<td><code>-v, --verbose</code></td>
<td>False</td>
<td>Print detailed progress</td>
</tr>
<tr>
<td><code>-q, --quiet</code></td>
<td>False</td>
<td>Suppress output except errors</td>
</tr>
</tbody>
</table>
<p><strong>Optimization options:</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--num-iterations</code></td>
<td>50</td>
<td>Maximum optimization iterations</td>
</tr>
<tr>
<td><code>--convergence-tol</code></td>
<td>0.01</td>
<td>Convergence tolerance</td>
</tr>
<tr>
<td><code>--convergence-window</code></td>
<td>3</td>
<td>Iterations to consider for convergence</td>
</tr>
<tr>
<td><code>--num-jackknife-blocks</code></td>
<td>100</td>
<td>Number of jackknife blocks for standard errors</td>
</tr>
<tr>
<td><code>--xtrace-num-samples</code></td>
<td>100</td>
<td>Samples for stochastic gradient estimation</td>
</tr>
<tr>
<td><code>--reset-trust-region</code></td>
<td>False</td>
<td>Reset trust region at each iteration</td>
</tr>
</tbody>
</table>
<p><strong>Processing options:</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--num-processes</code></td>
<td>auto</td>
<td>Number of parallel processes</td>
</tr>
<tr>
<td><code>--run-in-serial</code></td>
<td>False</td>
<td>Disable parallelization</td>
</tr>
<tr>
<td><code>--match-by-position</code></td>
<td>False</td>
<td>Match variants by position instead of RSID</td>
</tr>
<tr>
<td><code>--maximum-missingness</code></td>
<td>0.1</td>
<td>Maximum fraction of missing samples allowed</td>
</tr>
<tr>
<td><code>--max-chisq-threshold</code></td>
<td>None</td>
<td>Exclude blocks with chi-squared above threshold</td>
</tr>
</tbody>
</table>
<p><strong>Annotation options:</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--annotation-columns</code></td>
<td>all</td>
<td>Specific annotation columns to use</td>
</tr>
<tr>
<td><code>--binary-annotations-only</code></td>
<td>False</td>
<td>Only include 0/1 valued annotations</td>
</tr>
</tbody>
</table>
<p><strong>Gene annotation options</strong> (used with <code>-g, --gene-annot-dir</code>):</p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--gene-table</code></td>
<td>data/genes.tsv</td>
<td>Path to gene table TSV file</td>
</tr>
<tr>
<td><code>--nearest-weights</code></td>
<td>0.4,0.2,0.1,0.1,0.1,0.05,0.05</td>
<td>Weights for k-nearest genes</td>
</tr>
</tbody>
</table>
<p><strong>Advanced options:</strong></p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--surrogates</code></td>
<td>None</td>
<td>Path to precomputed surrogate markers HDF5</td>
</tr>
<tr>
<td><code>--initial-params</code></td>
<td>None</td>
<td>Initial parameter values (comma-separated)</td>
</tr>
<tr>
<td><code>--metadata</code></td>
<td>data/ldgms/metadata.csv</td>
<td>Path to LDGM metadata file</td>
</tr>
<tr>
<td><code>--score-test-filename</code></td>
<td>None</td>
<td>Output HDF5 for score test precomputation</td>
</tr>
<tr>
<td><code>--alt-output</code></td>
<td>False</td>
<td>Write separate files for heritability/enrichment/parameters</td>
</tr>
<tr>
<td><code>--no-save</code></td>
<td>False</td>
<td>Do not save results</td>
</tr>
</tbody>
</table>
<h3 id="gene-set-annotations">Gene Set Annotations<a class="headerlink" href="#gene-set-annotations" title="Permanent link">&para;</a></h3>
<p>GraphREML can use gene-level annotations from GMT files (Gene Matrix Transposed format). Gene sets are converted to variant-level annotations using a nearest-gene weighting scheme.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>reml<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="w">    </span>/path/to/sumstats.sumstats<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="w">    </span>output_prefix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="w">    </span>--gene-annot-dir<span class="w"> </span>/path/to/gmt/files/<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="w">    </span>--gene-table<span class="w"> </span>data/genes.tsv
</code></pre></div>
<p><strong>GMT file format:</strong>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>gene_set_name&lt;TAB&gt;description&lt;TAB&gt;gene1&lt;TAB&gt;gene2&lt;TAB&gt;...
</code></pre></div></p>
<p>Each gene set becomes a variant-level annotation where each variant receives a weighted score based on its proximity to genes in the set. The <code>--nearest-weights</code> option controls how the k-nearest genes contribute to each variant's score.</p>
<p>The gene table TSV must contain columns: <code>gene_id</code>, <code>gene_name</code>, <code>start</code>, <code>end</code>, <code>CHR</code>.</p>
<h3 id="surrogate-markers">Surrogate Markers<a class="headerlink" href="#surrogate-markers" title="Permanent link">&para;</a></h3>
<p>If variants are missing from your GWAS summary statistics, graphREML automatically assigns surrogate markers in high LD. This is common when using HapMap3 SNPs only (~1.1M SNPs).</p>
<p>To speed up repeated analyses, precompute surrogates:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>surrogates<span class="w"> </span>/path/to/sumstats.sumstats
</code></pre></div>
<p>Then pass the cached file with <code>--surrogates</code>. Surrogates don't need to match exactly and can be reused across similar sumstats files.</p>
<p>Precomputed surrogates for common reference panels are available for download from <a href="https://zenodo.org/records/8157131">Zenodo</a> (TODO: add direct link).</p>
<h3 id="parquet-multi-trait-files">Parquet Multi-Trait Files<a class="headerlink" href="#parquet-multi-trait-files" title="Permanent link">&para;</a></h3>
<p>When using parquet files with multiple traits:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="c1"># Process specific traits</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>reml<span class="w"> </span>sumstats.parquet<span class="w"> </span>output<span class="w"> </span>--name<span class="w"> </span>height,bmi<span class="w"> </span>...
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="c1"># Process all traits (omit --name)</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>reml<span class="w"> </span>sumstats.parquet<span class="w"> </span>output<span class="w"> </span>...
</code></pre></div>
<h2 id="blup-weights-blup">BLUP Weights (blup)<a class="headerlink" href="#blup-weights-blup" title="Permanent link">&para;</a></h2>
<p>Compute best linear unbiased predictor effect size estimates:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>blup<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="w">    </span>/path/to/sumstats.sumstats<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="w">    </span>output_prefix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="w">    </span>--heritability<span class="w"> </span><span class="m">0</span>.1
</code></pre></div>
<h2 id="ld-clumping-clump">LD Clumping (clump)<a class="headerlink" href="#ld-clumping-clump" title="Permanent link">&para;</a></h2>
<p>Perform p-value thresholding and LD-based pruning:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>clump<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="w">    </span>/path/to/sumstats.sumstats<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="w">    </span>output_prefix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="w">    </span>--rsq-threshold<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="w">    </span>--chisq-threshold<span class="w"> </span><span class="m">30</span>.0
</code></pre></div>
<h2 id="simulation-simulate">Simulation (simulate)<a class="headerlink" href="#simulation-simulate" title="Permanent link">&para;</a></h2>
<p>Simulate GWAS summary statistics:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>uv<span class="w"> </span>run<span class="w"> </span>graphld<span class="w"> </span>simulate<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="w">    </span>output_prefix<span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="w">    </span>--sample-size<span class="w"> </span><span class="m">10000</span><span class="w"> </span><span class="se">\</span>
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="w">    </span>--heritability<span class="w"> </span><span class="m">0</span>.5
</code></pre></div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["navigation.tabs", "navigation.sections", "navigation.expand", "navigation.top", "search.highlight", "content.code.copy"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": "1.2.0"}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>